---
title: "Results from rna_seq_kallisto1 pipeline"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
params:
  kal_folder: "kallisto"
  design: "test"
  p: "0.1"
  stat: "star_stats.tab"
  contrast_file: "contrast.tab"
---

```{r setup, include=FALSE}
library(Cairo)
knitr::opts_chunk$set(dev="CairoPNG")
knitr::opts_chunk$set(echo = FALSE)
library(dplyr)
library(ggplot2)
library(reshape2)
#library(DESeq2)
library(readr)
library(tximport)
#library(TxDb.Athaliana.BioMart.plantsmart28)
#source("deseq2_functions.R")
```

# {.tabset}

## Summery 

See also report.html 


## Used parameters

### Software and versions
```{r engine='bash', comment=''}
cat "modules.txt"
```


### Nextflow parameters
```{r engine='bash', comment=''}
cat "param.txt"
```

### SessionInfo from DEseq2 R session
```{r engine='bash', comment=''}
cat "sessionInfo_deseq2.txt"
```


## Sample info {#s_info}

```{r arguments, echo=FALSE}
contrast_file = params$contrast_file
kal_folder <- params$kal_folder
print(kal_folder)
s2c_file <- params$design
pvalue <- params$p
#sample_id <- dir(params$kal_folder)
#kal_dirs <- sapply(sample_id, function(id) file.path(params$kal_folder, id))
stat <- params$stat
s2c <- read.table(s2c_file, header = TRUE, stringsAsFactors=FALSE)
kable(s2c,caption ="samples file used")
```

## DESeq2 Results

### Description of analysis 
Differential expression analysis was carried out on the counts from [Kallisto](https://pachterlab.github.io/kallisto/about) using [DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). For the import of the transcript level data and summerazation into gene level data the R package [tximport](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html). 
**For information about versions used please refer to the ...**

A tab file with the resulting counts can be found here.. The samples are described under tab Sample info. 


### Pairs plot
<img src="pairs.png" alt="pairs plot">

Paired plot of all samples. Each dot is one gene, the x and y axes are the two samples, e.g. the plot .... The upper part of the plot shows the spearman correlation between the corresponding samples.


P-value threshold is set to `r I(pvalue)`

### PCA plot
<img src="pca.png" alt="pca plot">

PCA plot of all samples.


### Pair-wise DE analysis 

```{r, echo=FALSE,include=FALSE}
out = NULL
for ( i in dir("maplots")){
out = c(out, knit_child("ddeseq_contrast.Rmd"))
}
```


```{r,echo = FALSE, results='asis'}
cat(out)
```

Read more [here](./lists/contrast_group_ko_wt.csv) # It works!


## Alignment stats

```{r star_stat, message=FALSE}
star_stat=read.table(stat,sep="\t",header=T,stringsAsFactors = F)
star_stat = mutate(star_stat,sample=sapply(strsplit(ID,"Log"),"[[",1) )
star_stat = star_stat %>% select (-c(ID,X))
kable(star_stat, caption="Alignment stats (STAR)")
s2c <- mutate(s2c, name = paste(condition,sample,sep="_"))
s2c <-select(s2c, sample = run_accession, condition, name)
star_stat = inner_join(star_stat,s2c)
star_stat = star_stat %>% select (-c(sample,condition))
stat_mat = melt(star_stat, id = c("name"))
ggplot(stat_mat,aes(x=name,y=value,fill=name))+ geom_bar(stat="identity")+facet_wrap(~variable,scales = "free")+ guides(fill=FALSE)+
  theme(axis.title.x=element_blank(),axis.title.y=element_blank())
     
```
